sql/core/src/test/scala/org/apache/spark/sql/CTEInlineSuite.scala

```diff
@@ -703,6 +703,19 @@ abstract class CTEInlineSuiteBase
       checkErrorTableNotFound(e, "`tab_non_exists`", ExpectedContext("tab_non_exists", 83, 96))
     }
   }
+
+  test("SPARK-48307: not-inlined CTE references sibling") {
+    withTempView("t") {
+      val df = sql(
+        """
+          |WITH
+          |v1 AS (SELECT 1 col),
+          |v2 AS (SELECT col, rand() FROM v1)
+          |SELECT l.col FROM v2 l JOIN v2 r ON l.col = r.col
```

Comment:
before this PR, this query fails, because: `v1` can be inlined and removed from `WithCTE`. `v2` can't be inlined and kept in the `WithCTE`, but with its original form, which still contains a ref to `v1`.

Reply:
Is 

```java
java.util.NoSuchElementException: key not found: 0
	at scala.collection.MapOps.default(Map.scala:289)
	at scala.collection.MapOps.default$(Map.scala:288)
	at scala.collection.AbstractMap.default(Map.scala:420)
	at scala.collection.mutable.HashMap.apply(HashMap.scala:440)
	at org.apache.spark.sql.catalyst.optimizer.PushdownPredicatesAndPruneColumnsForCTEDef$.gatherPredicatesAndAttributes(PushdownPredicatesAndPruneColumnsForCTEDef.scala:74)
	at org.apache.spark.sql.catalyst.optimizer.PushdownPredicatesAndPruneColumnsForCTEDef$.$anonfun$gatherPredicatesAndAttributes$1(PushdownPredicatesAndPruneColumnsForCTEDef.scala:68)
	at scala.collection.immutable.Vector.foreach(Vector.scala:2124)
	at org.apache.spark.sql.catalyst.optimizer.PushdownPredicatesAndPruneColumnsForCTEDef$.gatherPredicatesAndAttributes(PushdownPredicatesAndPruneColumnsForCTEDef.scala:67)
```
the bug that you indicates?

Reply:
yes

Reply:
Can we put it into the PR description? 

BTW, SPARK-48307 has been identified as an improvement in the JIRA ticket. It now involves 3 PRs to fix bugs caused by the previous one. Should we revert the previous ones and squash them together for review?

Reply:
Added the error in the PR description

I'm not sure how reverting can help review, as the major change is in the first refactor PR. Followup PRs fix some flaws in the code but do not significantly change the code.

Reply:
Given that the LOC change here is small,  I agree with you with the AS-IS merging progress